cyw208xx PAwR updates

npal-cy opened the pull request:
Pushing updates which need for PAwR functionality on cyw208xx.

Pending work:
- update hal_infineon with new path for cyw208xx pwar fw blobs.

cc: @sreeramIfx

npal-cy committed:
driver/bluetooth / cyw208xx: update condition for select PAwR FW

Select PAwR if enabled following features:
- BT_PER_ADV_RSP
- BT_PER_ADV_SYNC_RSP

Remove dependency BT_AIROC_CUSTOM for AIROC_CUSTOM_FIRMWARE_HCD_BLOB

Signed-off-by: Nazar Palamar <nazar.palamar@infineon.com>

npal-cy committed:
Bluetooth: Samples: update response_cb in periodical_adv_rps

Remove print "Failed to receive response...", if receive
BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_RX_FAILED.

Problem: with cyw920829m2evk_02 device periodic_adv_rsp returns
multiple messages  “Failed to receive response….” in the terminal
log.

This happens because controller generates
HCI LE Periodic Advertising Response Report event for each
RspSlot. Because no RSP_DATA is observed in the OTA, which
is no scan device to transmit RSP_DATA, no RSP_DATA can be
received. Therefore, Data Status is 0xFF in all of these events,
which finally cause the messages  “Failed to receive response….”

Data_Status = 0xFF is expected because nothing is received
for that Rsp Slot.

Per spec, it’s not mandatory to generate these events,
So another vendors (e.g. Nordic)  might skip the HCI event
generation if Data_Status = 0xFF.

Signed-off-by: Nazar Palamar <nazar.palamar@infineon.com>

npal-cy committed:
soc: infineon/cyw20829: Update SYS_CLOCK_HW_CYCLES_PER_SEC

Update SYS_CLOCK_HW_CYCLES_PER_SEC to 96000000

Signed-off-by: Nazar Palamar <nazar.palamar@infineon.com>

npal-cy committed:
driver/bluetooth: Add setup of PUBLIC ADDR

Added setup of PUBLIC ADDR

NOTE: By default, the CYW208xx controller sets some
hard-coded static address.

To avoid address duplication, let's always override
the default address by using the HCI command
BT_HCI_VND_OP_SET_LOCAL_DEV_ADDR. So

1. when cyw208xx_setup gets BT_ADDR_ANY from the host,
   it will overwrite the default address, and the host
   will switch to using a random address (set in the hci_init
   function).

2. If user set the static address (by using bt_id_create)
   before bt_enable, cyw208xx_setup will set user defined
   static address.

Signed-off-by: Nazar Palamar <nazar.palamar@infineon.com>

jhedberg commented on drivers/bluetooth/hci/hci_ifx_cyw208xx.c:
```diff
@@ -185,6 +190,23 @@ static int cyw208xx_setup(const struct device *dev, const struct bt_hci_setup_pa
 	/* Waiting when BLE up after firmware launch */
 	cybt_platform_hci_wait_for_boot_fully_up(false);
+
+	/* Set public address */
+	buf = bt_hci_cmd_create(HCI_BRCM_SET_LOCAL_DEV_ADDR, BTM_SET_LOCAL_DEV_ADDR_LENGTH);
```
You shouldn't unconditionally do this, but only if the host actually has a public address to provide to the driver. If it doesn't have a public address the value will be `BT_ADDR_ANY`, so all of the public address related code should be behind something like this:
```c
	if (!bt_addr_eq(&params->public_addr, BT_ADDR_ANY)) {
		...
	}
```

npal-cy replied:
Hi @jhedberg 

Why i do this without conditions?

cyw208xx's controller by default set some hardcoded static address, so it will same for all cyw208xx devices.

I added set public address without condition. in case if user do not call bt_id_create before calling bt_enable() to define address, the  params->public_addr will be Zeros, so we will overwrite current hardcoded 20:82….:93 address with Zero.. As result bt_id_init will fail bt_setup_public_id_addr (because zero address) and set set_random_address.

So used doesn't set bt_id_create before bt_enable, we will use random address setup by bt_id_init (colled from hci_init)

Let me know if you see any issue here?

Regards,
Nazar

jhedberg replied:
Thanks for the clarification. Please add a code comment then explaining that this is intentional and what happens when you get `BT_ADDR_ANY` from the host.

npal-cy replied:
updated here https://github.com/zephyrproject-rtos/zephyr/pull/76857/commits/d5d8b5bcc01839802b0eef5fccf94edafe51a0c5. 

Is it clear now?

jhedberg replied:
> updated here [d5d8b5b](https://github.com/zephyrproject-rtos/zephyr/commit/d5d8b5bcc01839802b0eef5fccf94edafe51a0c5).
> 
> Is it clear now?

Yes, looks much better now. Thanks!

npal-cy committed:
Bluetooth: Samples: periodical_adv_rps: Added sync wait delay

Problem:
on cyw208xx devices i see sometime `Timed out while synchronizing`
error in samples periodical_adv_rps/periodical_sync_rps.

log:
<err> bt_scan: Could not lookup connection handle from PAST
<err> bt_scan: Unknown handle 0x0000 for periodic advertising report
Timed out while synchronizing
...

After investigation i can confirm, the sync was created successful,
but in application (periodical_sync_rsp) terminated the sync when the
connection is not valid.

periodical_adv_rps doesn't have any mechanism to wait for sync and after
gatt write (subevent, response_slot information) it immediately
performs disconnection.

Solution:
Added sync wait delay before make disconnection, to ensure sync
is established before disconnection.

Signed-off-by: Nazar Palamar <nazar.palamar@infineon.com>

jhedberg commented on drivers/bluetooth/hci/hci_ifx_cyw208xx.c:
```diff
@@ -165,11 +165,16 @@ static int cyw208xx_bt_firmware_download(const uint8_t *firmware_image, uint32_t
 
+#define HCI_BRCM_SET_LOCAL_DEV_ADDR     (0xFC01)
+#define BTM_SET_LOCAL_DEV_ADDR_LENGTH   6
+
+
```
minor: remove one of these empty lines (there should never be a need to have two consecutive ones)

jhedberg reviewed (dismissed):
@hermabe could you double check that the changes to `samples/bluetooth/periodic_adv_rsp` look ok?

npal-cy commented:
set draft due to pending work:  update hal_infineon with new path for cyw208xx pwar fw blobs.

hermabe commented on samples/bluetooth/periodic_adv_rsp/src/main.c:
```diff
@@ -369,6 +366,12 @@ int main(void)
+		 * used by controller) to ensure sync is established before
+		 * disconnection.
+		 */
+		k_sleep(K_MSEC(per_adv_params.interval_max * 2));
```
I guess this is ok. The samples do not handle failure well. A better solution might be for the sync sample to do the disconnect once it has synced to the pawr train, but that is not needed for this PR.

hermabe commented on samples/bluetooth/periodic_adv_rsp/src/main.c:
```diff
@@ -92,9 +92,6 @@ static void response_cb(struct bt_le_ext_adv *adv, struct bt_le_per_adv_response
 		bt_data_parse(buf, print_ad_field, NULL);
-	} else {
-		printk("Failed to receive response: subevent %d, slot %d\n", info->subevent,
-		       info->response_slot);
```
Since we listen to every response slot even if we dont expect any response this should be ok. Could potentially rephrase the print instead to not indicate an error, but that might get very spammy if the controller generates events for each response slot.

npal-cy replied:
yes it looks too spammy.

npal-cy committed:
west.yml: Update CYW20829 Bluetooth LE fw for PAwR support

Update CYW20829 Bluetooth LE fw for PAwR support
Commit#88d2529a3c5aee2e81947553bf6cbacb4671088c

Signed-off-by: Nazar Palamar <nazar.palamar@infineon.com>

npal-cy committed:
driver/bluetooth / cyw208xx: enable CYW20829_BT_FW_TX10DBM_POWER

Enable CYW20829_BT_FW_TX10DBM_POWER by default to get
best user experience experience out of the box.

Signed-off-by: Nazar Palamar <nazar.palamar@infineon.com>

npal-cy commented:
Hi @hermabe, @jhedberg, @jori-nordic this PR ready for merge. Could you please final review/approve?

Regards,
Nazar

hermabe reviewed (approved):
PAwR changes ok, not familiar with Infineon configs.
